libyul/CompilabilityChecker.cpp

```diff
@@ -0,0 +1,64 @@
+/*(
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Component that checks whether all variables are reachable on the stack.
+ */
+
+#include <libyul/CompilabilityChecker.h>
+
+#include <libyul/AsmAnalysis.h>
+#include <libyul/AsmAnalysisInfo.h>
+
+#include <libyul/backends/evm/EVMCodeTransform.h>
+#include <libyul/backends/evm/NoOutputAssembly.h>
+
+#include <liblangutil/EVMVersion.h>
+
+using namespace std;
+using namespace yul;
+using namespace dev;
+using namespace dev::solidity;
+
+std::map<YulString, int> CompilabilityChecker::run(std::shared_ptr<Dialect> _dialect, Block const& _ast)
+{
+	if (_dialect->flavour == AsmFlavour::Yul)
+		return {};
+
+	solAssert(_dialect->flavour == AsmFlavour::Strict, "");
+
+	EVMDialect const& evmDialect = dynamic_cast<EVMDialect const&>(*_dialect);
+
+	bool optimize = true;
+	yul::AsmAnalysisInfo analysisInfo =
+		yul::AsmAnalyzer::analyzeStrictAssertCorrect(_dialect, EVMVersion(), _ast);
+	NoOutputAssembly assembly;
+	CodeTransform transform(assembly, analysisInfo, _ast, evmDialect, optimize);
+	try
+	{
+		transform(_ast);
+	}
+	catch (StackTooDeepError const&)
```

Comment:
Hm... this component is later supposed to be used as metric for optimiser decisions? This is a bit nasty... checking for a property by checking whether exceptions are thrown... but if we extract the stack allocation out of the EVM bytecode generation later and can then get rid of this again, it might be ok as a hack...

Reply:
I agree. The exceptions are just there so that we do not accidentally take the bytecode from something that actually did not compile. Yes, we should single the stack allocation logic out of this at some point.